File: app/src/main/java/dev/ragnarok/filegallery/model/tags/TagFull.java
package dev.ragnarok.filegallery.model.tags;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.List;

import dev.ragnarok.filegallery.model.FileType;

public class TagFull implements Parcelable {

    public static final Creator<TagFull> CREATOR = new Creator<TagFull>() {
        @Override
        public TagFull createFromParcel(Parcel in) {
            return new TagFull(in);
        }

        @Override
        public TagFull[] newArray(int size) {
            return new TagFull[size];
        }
    };

    private String name;
    private List<TagDirEntry> dirs;

    public TagFull() {

    }

    protected TagFull(Parcel in) {
        name = in.readString();
        dirs = in.createTypedArrayList(TagDirEntry.CREATOR);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int flags) {
        parcel.writeString(name);
        parcel.writeTypedList(dirs);
    }

    public List<TagDirEntry> getDirs() {
        return dirs;
    }

    public TagFull setDirs(List<TagDirEntry> dirs) {
        this.dirs = dirs;
        return this;
    }

    public String getName() {
        return name;
    }

    public TagFull setName(String name) {
        this.name = name;
        return this;
    }

    public static class TagDirEntry implements Parcelable {

        public static final Creator<TagDirEntry> CREATOR = new Creator<TagDirEntry>() {
            @Override
            public TagDirEntry createFromParcel(Parcel in) {
                return new TagDirEntry(in);
            }

            @Override
            public TagDirEntry[] newArray(int size) {
                return new TagDirEntry[size];
            }
        };

        private String name;
        private String path;
        private @FileType
        int type = FileType.folder;

        public TagDirEntry() {

        }

        protected TagDirEntry(Parcel in) {
            name = in.readString();
            path = in.readString();
            type = in.readInt();
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel parcel, int flags) {
            parcel.writeString(name);
            parcel.writeString(path);
            parcel.writeInt(type);
        }

        public @FileType
        int getType() {
            return type;
        }

        public TagDirEntry setType(@FileType int type) {
            this.type = type;
            return this;
        }

        public String getName() {
            return name;
        }

        public TagDirEntry setName(String name) {
            this.name = name;
            return this;
        }

        public String getPath() {
            return path;
        }

        public TagDirEntry setPath(String path) {
            this.path = path;
            return this;
        }
    }
} 

File: app/src/main/kotlin/dev/ragnarok/filegallery/model/tags/TagFull.kt
package dev.ragnarok.filegallery.model.tags

import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.filegallery.model.FileType

class TagFull : Parcelable {
    var name: String? = null
        private set
    var dirs: List<TagDirEntry>? = null

    constructor()
    constructor(`in`: Parcel) {
        name = `in`.readString()
        dirs = `in`.createTypedArrayList(TagDirEntry.CREATOR)
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(name)
        parcel.writeTypedList(dirs)
    }

    fun setDirs(dirs: List<TagDirEntry>?): TagFull {
        this.dirs = dirs
        return this
    }

    fun setName(name: String?): TagFull {
        this.name = name
        return this
    }

    class TagDirEntry : Parcelable {
        var name: String? = null
            private set
        var path: String? = null
            private set

        @get:FileType
        @FileType
        var type = FileType.folder
            private set

        constructor()
        constructor(`in`: Parcel) {
            name = `in`.readString()
            path = `in`.readString()
            type = `in`.readInt()
        }

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(parcel: Parcel, flags: Int) {
            parcel.writeString(name)
            parcel.writeString(path)
            parcel.writeInt(type)
        }

        fun setType(@FileType type: Int): TagDirEntry {
            this.type = type
            return this
        }

        fun setName(name: String?): TagDirEntry {
            this.name = name
            return this
        }

        fun setPath(path: String?): TagDirEntry {
            this.path = path
            return this
        }

        companion object CREATOR : Parcelable.Creator<TagDirEntry> {
            override fun createFromParcel(parcel: Parcel): TagDirEntry {
                return TagDirEntry(parcel)
            }

            override fun newArray(size: Int): Array<TagDirEntry?> {
                return arrayOfNulls(size)
            }
        }
    }

    companion object CREATOR : Parcelable.Creator<TagFull> {
        override fun createFromParcel(parcel: Parcel): TagFull {
            return TagFull(parcel)
        }

        override fun newArray(size: Int): Array<TagFull?> {
            return arrayOfNulls(size)
        }
    }
}